libethcore/Miner.h

```diff
@@ -136,28 +163,112 @@ struct WorkingProgress
+        extendArray(m_lastUpdated, now, miner_index);
+        m_lastUpdated[miner_index] = now;
+    }
+    void rejected(unsigned miner_index)
+    {
+        extendArray(m_rejects, 0u, miner_index);
+        m_rejects[miner_index]++;
+        auto now = std::chrono::steady_clock::now();
+        extendArray(m_lastUpdated, now, miner_index);
+        m_lastUpdated[miner_index] = now;
+    }
+    void failed(unsigned miner_index)
+    {
+        extendArray(m_failures, 0u, miner_index);
+        m_failures[miner_index]++;
+        auto now = std::chrono::steady_clock::now();
+        extendArray(m_lastUpdated, now, miner_index);
+        m_lastUpdated[miner_index] = now;
+    }
+    void acceptedStale(unsigned miner_index)
+    {
+        extendArray(m_acceptedStales, 0u, miner_index);
+        m_acceptedStales[miner_index]++;
+        auto now = std::chrono::steady_clock::now();
+        extendArray(m_lastUpdated, now, miner_index);
+        m_lastUpdated[miner_index] = now;
+    }
 
-    void reset() { accepts = rejects = failures = acceptedStales = 0; }
+    unsigned getAccepts() const { return sumArray(m_accepts); }
+    unsigned getRejects() const { return sumArray(m_rejects); }
+    unsigned getFailures() const { return sumArray(m_failures); }
+    unsigned getAcceptedStales() const { return sumArray(m_acceptedStales); }
 
-    unsigned getAccepts() const { return accepts; }
-    unsigned getRejects() const { return rejects; }
-    unsigned getFailures() const { return failures; }
-    unsigned getAcceptedStales() const { return acceptedStales; }
+    unsigned getAccepts(unsigned miner_index) const
+    {
+        if (m_accepts.size() <= miner_index)
+            return 0;
+        return m_accepts[miner_index];
+    }
+    unsigned getRejects(unsigned miner_index) const
+    {
+        if (m_rejects.size() <= miner_index)
+            return 0;
+        return m_rejects[miner_index];
+    }
+    unsigned getFailures(unsigned miner_index) const
+    {
+        if (m_failures.size() <= miner_index)
+            return 0;
+        return m_failures[miner_index];
+    }
+    unsigned getAcceptedStales(unsigned miner_index) const
+    {
+        if (m_acceptedStales.size() <= miner_index)
+            return 0;
+        return m_acceptedStales[miner_index];
+    }
+    std::chrono::steady_clock::time_point getLastUpdated(unsigned miner_index) const
+    {
+        if (m_lastUpdated.size() <= miner_index)
+            return std::chrono::steady_clock::now();
+        return m_lastUpdated[miner_index];
+    }
 
 private:
-    unsigned accepts = 0;
-    unsigned rejects = 0;
-    unsigned failures = 0;
+    unsigned sumArray(const std::vector<unsigned>& array) const
+    {
+        unsigned r = 0;
+        for (size_t i = 0; i < array.size(); i++)
+            r += array[i];
+        return r;
+    }
+
+    template <typename T>
+    void extendArray(std::vector<T>& array, const T& initial_value, const size_t n) const
```

Comment:
Why does this method even exist? It is almost requivalent to .resize()!

Reply:
:+1: You're right ... I'm still learning c++ and getting knowledge of the std and boost library. So don't hesitate give me such hints.